src/Avalonia.Controls/ColorPicker/ColorSpectrum/ColorSpectrum.cs

```diff
@@ -0,0 +1,1572 @@
+        // associated with the last call to CreateBitmapsAndColorMap(),
+        // in order to function properly while the asynchronous bitmap creation
+        // is in progress.
+        private ColorSpectrumShape _shapeFromLastBitmapCreation = ColorSpectrumShape.Box;
+        private ColorSpectrumChannels _componentsFromLastBitmapCreation = ColorSpectrumChannels.HueSaturation;
+        private double _imageWidthFromLastBitmapCreation = 0.0;
+        private double _imageHeightFromLastBitmapCreation = 0.0;
+        private int _minHueFromLastBitmapCreation = 0;
+        private int _maxHueFromLastBitmapCreation = 0;
+        private int _minSaturationFromLastBitmapCreation = 0;
+        private int _maxSaturationFromLastBitmapCreation = 0;
+        private int _minValueFromLastBitmapCreation = 0;
+        private int _maxValueFromLastBitmapCreation = 0;
+
+        private Color _oldColor = Color.FromArgb(255, 255, 255, 255);
+        private HsvColor _oldHsvColor = HsvColor.FromAhsv(0.0f, 0.0f, 1.0f, 1.0f);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ColorSpectrum"/> class.
+        /// </summary>
+        public ColorSpectrum()
+        {
+            _shapeFromLastBitmapCreation = Shape;
+            _componentsFromLastBitmapCreation = Channels;
+            _imageWidthFromLastBitmapCreation = 0;
+            _imageHeightFromLastBitmapCreation = 0;
+            _minHueFromLastBitmapCreation = MinHue;
+            _maxHueFromLastBitmapCreation = MaxHue;
+            _minSaturationFromLastBitmapCreation = MinSaturation;
+            _maxSaturationFromLastBitmapCreation = MaxSaturation;
+            _minValueFromLastBitmapCreation = MinValue;
+            _maxValueFromLastBitmapCreation = MaxValue;
+        }
+
+        /// <inheritdoc/>
+        protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
+        {
+            base.OnApplyTemplate(e);
+
+            UnregisterEvents();
+
+            _colorNameToolTip = e.NameScope.Find<ToolTip>("PART_ColorNameToolTip");
+            _inputTarget = e.NameScope.Find<Canvas>("PART_InputTarget");
+            _layoutRoot = e.NameScope.Find<Grid>("PART_LayoutRoot");
+            _selectionEllipsePanel = e.NameScope.Find<Panel>("PART_SelectionEllipsePanel");
+            _sizingGrid = e.NameScope.Find<Grid>("PART_SizingGrid");
+            _spectrumEllipse = e.NameScope.Find<Ellipse>("PART_SpectrumEllipse");
+            _spectrumRectangle = e.NameScope.Find<Rectangle>("PART_SpectrumRectangle");
+            _spectrumOverlayEllipse = e.NameScope.Find<Ellipse>("PART_SpectrumOverlayEllipse");
+            _spectrumOverlayRectangle = e.NameScope.Find<Rectangle>("PART_SpectrumOverlayRectangle");
+
+            if (_layoutRoot != null)
+            {
+                _layoutRootDisposable = _layoutRoot.GetObservable(BoundsProperty).Subscribe(_ => 
+                {
+                    CreateBitmapsAndColorMap();
+                });
```

Comment:
I wonder if it should be subscribed in AttachedToVisualTree and unsubed in DettachedToVisualTree. 
As well as other events in UnregisterEvents. That would be usual pattern.

Though there should be any problems/memory leak with current solution, because we subscribe only on current template's parts, which will die together with the control.

Reply:
These two property change subscriptions are just left where they were in the C++ version. I realize it is better to put these in Loaded/Unloaded (and AttachedToVisualTree/DetachedToVisualTree are a direct equivalents). So I can move these to make it more in line with what we would normally see.

Reply:
So I just tested this and it doesn't work. `OnAttachedToVisualTree()` is called BEFORE `OnApplyTemplate()`. However, we have to have references to the controls which doesn't exist until `OnApplyTemplate()` is called.

This is a good reason to add a `Loaded` and `Unloaded` event from UWP. Loaded is called AFTER `OnApplyTemplate()` and just before the control is visible -- literally the last step.

Reply:
If your intent is to intercept the change, just override the OnPropertyChanged method

Reply:
> If your intent is to intercept the change, just override the OnPropertyChanged method

Well, generally yes and that is done elsewhere; however, this property isn't in the control itself -- it is is one of the template parts so wouldn't invoke OnPropertyChanged.

That said, this area was basically a direct port and I suppose I didn't think about it critically. If the layout root changes, the control bounds itself should change as well. The layout root is stretched to the control. To be perfectly safe the sizing grid or one of the spectrum rectangles/ellipses should have been used anyway.

I also just noticed SizingGrid sits directly inside LayoutRoot... why UWP did this I don't know and it's completely unnecessary it seems.

Reply:
I really don't know why UWP did this:

https://github.com/microsoft/microsoft-ui-xaml/blob/2e53d0bd1e5f786bf02d402873d0c78407524339/dev/ColorPicker/ColorSpectrum.cpp#L61-L86

Anyway, both SizeChanged (Bounds) and FlowDirection can come from the control itself as far as I understand things. It doesn't appear necessary for accessibility either. So good catch @workgroupengineering. I will make changes to use OnPropertyChanged.

I am a little more hesitant about FlowDirection but my understanding is that it should be inherited even in template parts. So will change things there too. Just something to keep in mind if FlowDirection seems incorrect later on.

Reply:
Code updated and these two observables were completely removed. Now the `BoundsProperty` and `FlowDirectionProperty` of the control itself are used. As stated, I can't see anything wrong with doing this.

Reply:
@maxkatz6 @workgroupengineering If you don't see something I missed by changing the code I'm going to consider this one resolved.

Reply:
Well, this change somehow broke theme switching so I had to revert it. Back to watching for changes on template parts instead of the control itself.